### PR TITLE
Add controller_group script that allows switching groups easily

### PR DESCRIPTION
**Pull request by yli-cpr**
The groups are read from ROS parameter "/controller_groups".

This is for the scenario that there can be multiple set of controllers that talk to same hardware interfaces but for different purposes. Without this script, one will have to stop all the production controllers and start a special set of controllers one by one. This script makes switching from a set to another much easier.

### CODE REVIEW

File: controller_manager/scripts/controller_group
Comment (mikepurvis) [inline controller_manager/scripts/controller_group]: Initializing a node may actually not be necessary, since you're only calling services and not interacting with topics at all.

File: controller_manager_tests/test/controller_manager_scripts.test
Comment (mikepurvis) [inline controller_manager_tests/test/controller_manager_scripts.test:17]: Should there be some kind of failure case here, like a group where two of the controllers try to claim the same resource? Does the new tool fail gracefully under those circumstances?

File: controller_manager/scripts/controller_group
Comment (mikepurvis) [inline controller_manager/scripts/controller_group]: Since this class has no state, IMO it's superfluous. Consider moving the methods up to module level (eg `list_groups`, `spawn_group`, etc).

File: controller_manager/scripts/controller_group
Comment (mikepurvis) [inline controller_manager/scripts/controller_group]: IMO this is fine for now, but down the road it might be desirable to have this be configurable somehow, possibly an arg which allows passing the name of the group param?

File: controller_manager/scripts/controller_group
Comment (mikepurvis) [inline controller_manager/scripts/controller_group]: For py3 friendliness, prefer just `items`. For such a small list, the savings from the iterator are extremely small anyway.

**Review by mikepurvis — CHANGES_REQUESTED** [review]
Looks great, mostly just piddly stuff.

File: controller_manager/scripts/controller_group
Comment (yli-cpr) [reply controller_manager/scripts/controller_group]: Added for rospy.myargv() but seems not necessary.

File: controller_manager_tests/test/controller_manager_scripts.test
Comment (yli-cpr) [reply controller_manager_tests/test/controller_manager_scripts.test:17]: I am trying a failure case. The script is not that smart to verify this error, but it will print errors.

File: controller_manager/scripts/controller_group
Comment (yli-cpr) [reply controller_manager/scripts/controller_group]: sure

File: controller_manager/scripts/controller_group
Comment (yli-cpr) [reply controller_manager/scripts/controller_group]: Sure. Also I need sorted() for python 2 to print stable output

File: controller_manager/scripts/controller_group
Comment (yli-cpr) [reply controller_manager/scripts/controller_group]: Yeah. Also it can be a namespace? CombineRobotHW is loading `robot_hardware` from the node's namespace. I don't have much time available, so I would leave it for now.

File: controller_manager/scripts/controller_group
Comment (mikepurvis) [reply controller_manager/scripts/controller_group]: That's a good point— this definitely does still need to work when run from a launch file.

File: controller_manager_tests/test/controller_manager_scripts.test
Comment (mikepurvis) [reply controller_manager_tests/test/controller_manager_scripts.test:17]: The underlying SwitchController service call should return a `False`:

http://docs.ros.org/lunar/api/controller_manager_msgs/html/srv/SwitchController.html

So I think it's mostly just on the script here to detect that and bubble some kind of error message to the user.

File: controller_manager/scripts/controller_group
Comment (yli-cpr) [reply controller_manager/scripts/controller_group]: do you mean init_node is still needed?

File: controller_manager/scripts/controller_group
Comment (mikepurvis) [reply controller_manager/scripts/controller_group]: Using `ROS_NAMESPACE` could be nice because then you can do it once in your terminal and the ergonomics of the CLI usage are otherwise preserved (not needing an extra flag all the time). And you can get that pretty much for free I think by simply removing the leading slash from the param here.

Anyway, apart from that there isn't a single obvious way forward with this; since we don't have an immediate need, I'd just leave it.

File: controller_manager/scripts/controller_group
Comment (mikepurvis) [reply controller_manager/scripts/controller_group]: I don't think it should be; `myargv` will have you covered— was more just acknowledging that that use case be considered.

File: controller_manager/scripts/controller_group
Comment (yli-cpr) [inline controller_manager/scripts/controller_group]: Now using relative parameters, so that it respects ROS_NAMESPACE or roslaunch remapping. Confirmed it works.

File: controller_manager/CMakeLists.txt
Comment (mikepurvis) [inline controller_manager/CMakeLists.txt]: Alpha ordering here.

File: controller_manager/scripts/controller_group
Comment (mikepurvis) [inline controller_manager/scripts/controller_group]: I know you copied this from `controller_manager`, but please use a full BSD header:

https://github.com/yli-cpr/ros_control/blob/f7c91ad13da18a3241990e3eeca2b34b092db45d/controller_manager/scripts/spawner#L1-L33

File: controller_manager/scripts/controller_group
Comment (mikepurvis) [reply controller_manager/scripts/controller_group]: Consider adding a comment explaining this to future readers/devs.

File: controller_manager_tests/test/controller_manager_scripts.py
Comment (mikepurvis) [inline controller_manager_tests/test/controller_manager_scripts.py]: Shouldn't the switching happen atomically? Like, stop old and start new all in one go? That's one of the ideas of ros_control, is that even when you hot-switch controllers, it's possible for the system to never be uncontrolled.

So the output I'd expect here would be that the old controllers continue to run because the new config had a conflict and therefore wasn't loaded.

Comment (yli-cpr) [reply controller_manager_tests/test/controller_manager_scripts.py]: I didn't use switch_controllers, and the reason is... probably silly. I am doing it with switch_controllers now.

File: controller_manager/scripts/controller_group
Comment (yli-cpr) [reply controller_manager/scripts/controller_group]: sure

**Comment by yli-cpr** [comment]
got it. fixing the error.

File: controller_manager_tests/test/controller_manager_scripts.test
Comment (yli-cpr) [reply controller_manager_tests/test/controller_manager_scripts.test:17]: Now it uses switch_controllers service which will print this
```
[\ERROR] [\/dummy_app] [\1525462144.701348031]: \Could not switch controllers, due to resource conflict
```
The script will also print an error based on the false return value.

File: controller_manager/scripts/controller_group
Comment (mikepurvis) [inline controller_manager/scripts/controller_group]: Why is there a conditional here? If a controller that's needed couldn't be loaded, the whole operation should fail, right?

Comment (yli-cpr) [reply controller_manager/scripts/controller_group]: I don't see switch_controller service loads the controller. So I have to load those controllers that haven't been loaded. Yeah. I should check `load_controller()`'s return value.

The second `if` is because switch_controllers service will fail if I try to start a controller that's already running, because we are using `Strict`

Comment (mikepurvis) [reply controller_manager/scripts/controller_group]: Oh, got it— so you're figuring out the stops/starts required to get from the current state to the desired state, as opposed to just stop everything and start <desired>.

File: controller_manager_tests/test/controller_manager_scripts.test
Comment (mikepurvis) [reply controller_manager_tests/test/controller_manager_scripts.test:17]: 👍

**Review by mikepurvis — APPROVED** [review]

**Comment by mikepurvis** [comment]
This looks great to me. Ping @efernandez and @bmagyar for their thoughts.

File: controller_manager/scripts/controller_group
Comment (efernandez) [inline controller_manager/scripts/controller_group]: If groups is a `dict` you can also do `for controllers in groups.itervalues():`, which I think it's preferred over `values()`.

Then I'd update the comment below, b/c there won't be any `list`, to:
` # Add to a set to remove duplicates`

Or remove the comment from here and move it before `all_controllers = set()` and say `# Remove duplicates`.
I personally prefer this last option.

File: controller_manager/scripts/controller_group
Comment (efernandez) [inline controller_manager/scripts/controller_group]: I think `all_controllers.add(controllers)` is easier to read, at least for me.
I'm not familiar with the syntax you're using.

File: controller_manager/scripts/controller_group
Comment (efernandez) [inline controller_manager/scripts/controller_group]: Shouldn't we pass a `timeout=N`, where I think `N` could be 5s or so. Otherwise this could get block here and the user won't know what's going on.

If you add the timeout you need a try - except block catching `ROSException`.

File: controller_manager/scripts/controller_group
Comment (efernandez) [inline controller_manager/scripts/controller_group]: This is the same as `resp = s(ListControllerRequest())` IIRC.

You also need a try - except block for this, in case the call fails for any of these reasons:
``` python
@raise TypeError: if request is not of the valid type (Message)
@raise ServiceException: if communication with remote service fails
@raise ROSInterruptException: if node shutdown (e.g. ctrl-C) interrupts service call
@raise ROSSerializationException: If unable to serialize
```

In this case, it's probably not a big deal, but I think `ServiceException` and `ROSInterruptException` should be caught. (or their base class, if any).

File: controller_manager/scripts/controller_group
Comment (efernandez) [inline controller_manager/scripts/controller_group:71]: Why do you use `print` here instead of `rospy.loginfo`?

In other parts of the code you use `rospy.logerr`, for instance.

File: controller_manager/scripts/controller_group
Comment (efernandez) [inline controller_manager/scripts/controller_group:135]: Uhm, I thought you had to `rospy.init_node('controller_group', sys.argv, anonymous=True)` to call services.
I think it's better to do so: the time will be initialized, and it'd allow to do remappings, although it's something we probably don't need here.

Another thing, that I think it's common, it's to have a try - except block to check for `KeyError`, so we detect when the user Ctrl+C the script.

File: controller_manager/src/controller_manager/controller_manager_interface.py
Comment (efernandez) [inline controller_manager/src/controller_manager/controller_manager_interface.py]: Or `start_controllers([name])`

File: controller_manager/src/controller_manager/controller_manager_interface.py
Comment (efernandez) [inline controller_manager/src/controller_manager/controller_manager_interface.py]: Or `stop_controllers([name])`

File: controller_manager/src/controller_manager/controller_manager_interface.py
Comment (efernandez) [inline controller_manager/src/controller_manager/controller_manager_interface.py]: Not really from this PR, but you could add parenthesis to `print`. Here and below

File: controller_manager_tests/test/controller_manager_scripts.py
Comment (efernandez) [inline controller_manager_tests/test/controller_manager_scripts.py]: Could you also move group1 to its own line, please? It'd be easier to read.

**Review by efernandez — CHANGES_REQUESTED** [review]
LGTM, only minor comments.

File: controller_manager/src/controller_manager/controller_manager_interface.py
Comment (bmagyar) [reply controller_manager/src/controller_manager/controller_manager_interface.py]: Make sure to put this in a separate commit please, I'd cherry pick it to kinetic

File: controller_manager/scripts/controller_group
Comment (bmagyar) [reply controller_manager/scripts/controller_group]: Or let it err out with the exception, although less pretty, that's not bad either.

File: controller_manager/scripts/controller_group
Comment (yli-cpr) [reply controller_manager/scripts/controller_group]: The problem is itervalues() is not supported by Python 3. Also the performance gain is trivial for this use case. Will move the comment.

File: controller_manager/scripts/controller_group
Comment (yli-cpr) [reply controller_manager/scripts/controller_group]: `controllers` is a list of objects. I can change it to `all_controllers.union(set(controllers))` or add another loop.

File: controller_manager/scripts/controller_group
Comment (yli-cpr) [reply controller_manager/scripts/controller_group]: Was following controller_manager_interface. I will add the a timeout and let it crash with the exception.

File: controller_manager/scripts/controller_group
Comment (yli-cpr) [reply controller_manager/scripts/controller_group]: When they are not caught, the exception will be logged. This script isn't supposed to keep running. What value does catching them give us?

File: controller_manager/scripts/controller_group
Comment (yli-cpr) [reply controller_manager/scripts/controller_group:71]: Because in this case it is supposed to print the list to stdout, like printing usages. It shouldn't go through any logger IMHO.

File: controller_manager/scripts/controller_group
Comment (yli-cpr) [reply controller_manager/scripts/controller_group:135]: I did that for a script that uses Twisted to stop reactor. I think it should be done when necessary. If I add it for this script, I don't know what to do with with exception except printing something. Would leave it to python default behavior.

File: controller_manager/src/controller_manager/controller_manager_interface.py
Comment (yli-cpr) [reply controller_manager/src/controller_manager/controller_manager_interface.py]: Was trying to follow the style of `controller_manager_interface.py`. Also `_controllers` doesn't give more information because the function is called `start_stop_controllers()`

File: controller_manager/src/controller_manager/controller_manager_interface.py
Comment (yli-cpr) [reply controller_manager/src/controller_manager/controller_manager_interface.py]: @bmagyar Do you mean `print(msg)` style shouldn't be in kinetic?

File: controller_manager_tests/test/controller_manager_scripts.py
Comment (yli-cpr) [reply controller_manager_tests/test/controller_manager_scripts.py]: Sure

File: controller_manager/scripts/controller_group
Comment (efernandez) [reply controller_manager/scripts/controller_group]: I think I prefer `all_controllers.union(set(controllers))`, but I leave it up to you.

File: controller_manager/scripts/controller_group
Comment (efernandez) [reply controller_manager/scripts/controller_group]: Ok, nvm.

Sounds good.

File: controller_manager/scripts/controller_group
Comment (efernandez) [reply controller_manager/scripts/controller_group:71]: I see, makes sense.

File: controller_manager/scripts/controller_group
Comment (efernandez) [reply controller_manager/scripts/controller_group:135]: Ok. As long as the user can Ctrl+C to stop the script if needed. It looks good to me.

File: controller_manager/src/controller_manager/controller_manager_interface.py
Comment (efernandez) [reply controller_manager/src/controller_manager/controller_manager_interface.py]: The reason I'd call `start_controllers([name])` is b/c if in the future sth is added there, it's better to have a single entry point. I think we'll always handle a single element the same way as a list of a single element.

File: controller_manager/scripts/controller_group
Comment (efernandez) [reply controller_manager/scripts/controller_group]: I think you're right, nvm.

File: controller_manager/src/controller_manager/controller_manager_interface.py
Comment (efernandez) [reply controller_manager/src/controller_manager/controller_manager_interface.py]: I think he wants to cherry-pick only that style/deprecation change into `kinetic-devel`.
This PR is targeting `modelodic-devel`, btw.

File: controller_manager/scripts/controller_group
Comment (yli-cpr) [reply controller_manager/scripts/controller_group]: will do `union`

File: controller_manager/src/controller_manager/controller_manager_interface.py
Comment (yli-cpr) [reply controller_manager/src/controller_manager/controller_manager_interface.py]: Got it. Sure.

File: controller_manager/src/controller_manager/controller_manager_interface.py
Comment (yli-cpr) [reply controller_manager/src/controller_manager/controller_manager_interface.py]: I see.

File: controller_manager/scripts/controller_group
Comment (yli-cpr) [reply controller_manager/scripts/controller_group]: actually there is an `update` method for `set` which accepts any iterable as input. will use that.

**Review by bmagyar — CHANGES_REQUESTED** [review]
Great work! I have one request before putting a green light on this: Please add documentation here: http://wiki.ros.org/controller_manager

**Comment by yli-cpr** [comment]
@bmagyar The wiki page has been updated.
Ping @efernandez for another look.

**Review by bmagyar — APPROVED** [review]

**Comment by bmagyar** [comment]
@efernandez feel free to merge if you greenlight it too!

**Review by efernandez — APPROVED** [review]

**Comment by mikepurvis** [comment]
Awesome, thanks for your work on this, @yli-cpr! A possible follow-on at some point might be to add a `controller_group status` verb or similar, which would examine the currently running controllers and tell you which (if any) group is currently active.

**Comment by yli-cpr** [comment]
That's good idea. Will add it to todo list

**Pull request merged**
